### PR TITLE
`onnxruntime/modeling_ort.py` refactor, part 1

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This PR aims at simplifying and cleaning the `modeling_ort.py` and related files. This will be done in multiple PRs to make it easier to review.

- [x] Remove `__init__` that are not needed
- [x] Make `prepare_io_binding` generic enough so that it is not needed to implement it for every subclass

Left remaining for following PRs:

- [ ] Check dynamic axes for the inputs and outputs of exported models with `exporters.onnx`
   - [ ] Add tests to find problematic models
   - [ ]  `M2M100SinusoidalPositionalEmbedding` makes shape inference fail and creates a dynamic axis for `last_hidden_state` of the encoder, but it is actually static, fix it.
- [ ] Check past key values shape definition in `exporters.onnx` and other related inputs / outputs dynamic axes in decoder and seq2seq models
- [ ] Check which inputs are given to the model according to the `InferenseSession` (for `prepare_io_binding` and regular forward)
- [ ] Fix inference with Whisper, which fails under this PR
- [ ] Fix Marian `.generate`

### CODE REVIEW

**Review by fxmarty — APPROVED** [review]
LGTM, can you make sure the GPU tests pass locally?

File: optimum/onnxruntime/modeling_ort.py
Comment (fxmarty) [inline optimum/onnxruntime/modeling_ort.py:620]: Why not use `self.model_inputs` and `self.model_outputs` here?

File: optimum/onnxruntime/modeling_ort.py
Comment (JingyaHuang) [inline optimum/onnxruntime/modeling_ort.py:622]: How are you going to deal with shapes on dynamic axis? The definition could change according to task, got to be careful.

File: optimum/onnxruntime/modeling_ort.py
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_ort.py:269]: This is actually very important, because the way it was done previously did not handle this case:

```python
my_model = ORTModel(..., providers=["CPUExecutionProvider"]).to("cuda")
```

Before this, we would not use IO binding because `self.use_io_binding` was set at initialization.

File: optimum/onnxruntime/modeling_ort.py
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_ort.py:647]: Waiting to remove, to check if this is actually needed.

File: optimum/onnxruntime/modeling_ort.py
Comment (michaelbenayoun) [reply optimum/onnxruntime/modeling_ort.py]: We are getting them when possible, otherwise we can specify them via the `known_output_shapes` argument.

File: optimum/exporters/onnx/model_configs.py
Comment (fxmarty) [inline optimum/exporters/onnx/model_configs.py:409]: Why is this necessary?

File: optimum/exporters/onnx/model_configs.py
Comment (fxmarty) [inline optimum/exporters/onnx/model_configs.py:914]: Why is this necessary?

File: optimum/onnxruntime/base.py
Comment (fxmarty) [inline optimum/onnxruntime/base.py:36]: I'm not a huge fan of passing a parent class to its children; it feels like it can be bug prone. Couldn't we simply pass the relevant attributes, and possibly update them when needed (e.g. device)

Also, not sure if the parent model will actually have a `.config` attribute, e.g. see https://huggingface.co/stabilityai/stable-diffusion-2-1/tree/main . But we can leave it to an other PR to make it more general.

File: optimum/onnxruntime/base.py
Comment (fxmarty) [inline optimum/onnxruntime/base.py:73]: This basically makes the attention_mask optional?

File: optimum/onnxruntime/base.py
Comment (fxmarty) [inline optimum/onnxruntime/base.py:140]: may `self.expected_key_symbolic_shape` not be None here?

File: optimum/onnxruntime/base.py
Comment (fxmarty) [inline optimum/onnxruntime/base.py:314]: is this also making those inputs facultative?

File: optimum/onnxruntime/base.py
Comment (fxmarty) [inline optimum/onnxruntime/base.py:306]: Why is this needed ?

File: optimum/exporters/onnx/convert.py
Comment (fxmarty) [inline optimum/exporters/onnx/convert.py:54]: These are actually not used in `__main__.py` of the ONNX export, is it normal?

File: tests/onnxruntime/test_modeling.py
Comment (fxmarty) [inline tests/onnxruntime/test_modeling.py:2226]: to remove?

File: optimum/onnxruntime/base.py
Comment (fxmarty) [inline optimum/onnxruntime/base.py:164]:
```suggestion
            half_shape[0] = batch_size * num_attention_heads
            del half_shape[1]
```

File: optimum/exporters/onnx/model_configs.py
Comment (fxmarty) [inline optimum/exporters/onnx/model_configs.py:912]: test ordereddict update

File: optimum/utils/normalized_config.py
Comment (michaelbenayoun) [reply optimum/utils/normalized_config.py:163]: Handled.

**Review by JingyaHuang — COMMENTED** [review]
LGTM, just left two nits.

File: optimum/exporters/onnx/base.py
Comment (mht-sharma) [inline optimum/exporters/onnx/base.py:126]: small nit: Since this is a refactor, could you also make the following list alphabetical? I guess the audio keys are only misaligned.

File: optimum/exporters/onnx/base.py
Comment (mht-sharma) [inline optimum/exporters/onnx/base.py:596]:
```suggestion
            if self._behavior is ConfigBehavior.ENCODER or "encoder" in name:
                sequence_name = "encoder_sequence_length"
            else:
```

File: optimum/exporters/onnx/base.py
Comment (mht-sharma) [inline optimum/exporters/onnx/base.py:606]: Does this happen also when we do not put the axis_id in dict?

File: optimum/exporters/onnx/config.py
Comment (mht-sharma) [inline optimum/exporters/onnx/config.py:110]: It still is `encoder_outputs` when the `def inputs` mean `torch inputs for export`, but would mean `encoder_hidden_state` if considered ONNX inputs.

For me it's the former and the naming is correct. Maybe a comment would help to avoid confusion?

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (mht-sharma) [inline optimum/onnxruntime/modeling_seq2seq.py:359]: How are you handling this case?

File: optimum/exporters/onnx/base.py
Comment (michaelbenayoun) [reply optimum/exporters/onnx/base.py:606]: Yes, basically I do that because ONNX infers that the axis is dynamic but it is not. I will look into this problem, which is not limited to this case, on another PR.

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (michaelbenayoun) [reply optimum/onnxruntime/modeling_seq2seq.py:359]: It is not handled. But as soon as this is merged, will work on a solution for that. The idea is to keep this PR as small as possible.

File: optimum/exporters/onnx/convert.py
Comment (michaelbenayoun) [reply optimum/exporters/onnx/convert.py:54]: Yes, will be used in a PR coming this week.

File: optimum/onnxruntime/utils.py
Comment (michaelbenayoun) [reply optimum/onnxruntime/utils.py:227]: About your first question, I do not think so because in the `to` method we update the providers, which can be blocking to have `use_io_binding`. But we actually dont want to use io binding when the initialization was done so that `self._use_io_binding = False`.

Comment (michaelbenayoun) [reply optimum/onnxruntime/utils.py:227]: I think you're right, removed it.

**Pull request merged**
